### PR TITLE
Fix JSON exception in malformed IDF

### PR DESCRIPTION
**Pull request by mbadams5**
Pull request overview
---------------------
This fixes part of #7806 

This PR addresses the json.exception.type_error.305 generated in #7806, but it does not address the ExpandObjects bug.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Review by Myoldmopar — APPROVED** [review]
Certainly an improvement here.  I'll test locally but it should be good to go since CI is happy with the code and unit test.

**Comment by Myoldmopar** [comment]
Built locally with develop pulled in, no problem.  Unit tests run, no problem.  I did try to hack up a defect file on my own to see what I could do.  I took WindACRHControl.idf, went to line 1754, and changed the trailing semicolon on the DistrictHeating object to be a comma.  Kinda like the unit test -- I know the field order is different and what-not.  I ran it, and it did give an informative error -- sorta.  The error file looks like this:

```
Program Version,EnergyPlus, Version 9.3.0-0ec1c8281b, YMD=2020.03.02 19:53,
   ** Severe  ** Line: 1756 Index: 21 - Error parsing "DistrictHeating". Error in following line.
   ** Severe  ** ~~~     HW Circ Pump,            !- Name
    HW Supply Inlet Node,    !- Inlet Node Name
    HW Pump Outlet Node,     !- Outlet Node Name
    0.005,                   !- Design Maximum Flow Rate {m3/s}
    300000,                  !- Design Pump Head {Pa}
    2250,                    !- Design Power Consumption {W}
    0.87,                    !- Motor Efficiency
    0.0,                     !- Fraction of Motor Inefficiencies to Fluid Stream
    0,                       !- Coefficient 1 of the Part Load Performance Curve
    1,                       !- Coefficient 2 of the Part Load Performance Curve
    0,                       !- Coefficient 3 of the Part Load Performance Curve
    0,                       !- Coefficient 4 of the Part Load Performance Curve
    0,                       !- Design Minimum Flow Rate {m3/s}
    INTERMITTENT;            !- Pump Control Type

  Coil:Cooling:DX:SingleSpeed,
    Zone1WindACDXCoil,       !- Name
    CoolingCoilAvailSched,   !- Availability Schedule Name
    10548,                   !- Gross Rated Total Cooling Capacity {W}
    0.75,                    !- Gross Rated Sensible Heat Ratio
    3.0,                     !- Gross Rated Cooling COP {W/W}
    0.6,                     !- Rated Air Flow Rate {m3/s}
    ,                        !- Rated Evaporator Fan Power Per Volume Flow Rate {W/(m3/s)}
    Zone1WindACFanOutletNode,!- Air Inlet Node Name
    Zone1WindACAirOutletNode,!- Air Outlet Node Name
    WindACCoolCapFT,         !- Total Cooling Capacity Function of Temperature Curve Name
    WindACCoolCapFFF,        !- Total Cooling Capacity Function of Flow Fraction Curve Name
    WindACEIRFT,             !- Energy Input Ratio Function of Temperature Curve Name
    WindACEIRFFF,            !- Energy Input Ratio Function of Flow Fraction Curve Name

<snip because the whole rest of the IDF is in here>

  Output:Surfaces:Drawing,dxf;

  OutputControl:Table:Style,
    HTML;                    !- Column Separator

  Output:Table:SummaryReports,
    AllSummary;              !- Report 1 Name


   ** Severe  ** Line: 1757 Index: 16 - "HW Circ Pump" is not a valid Object Type.
   **  Fatal  ** Errors occurred on processing input file. Preceding condition(s) cause termination.
   ...Summary of Errors that led to program termination:
   ..... Reference severe error count=3
   ..... Last severe error=Line: 1757 Index: 16 - "HW Circ Pump" is not a valid Object Type.
   ************* Warning:  Node connection errors not checked - most system input has not been read (see previous warning).
   ************* Fatal error -- final processing.  Program exited before simulations began.  See previous error messages.
   ************* EnergyPlus Warmup Error Summary. During Warmup: 0 Warning; 0 Severe Errors.
   ************* EnergyPlus Sizing Error Summary. During Sizing: 0 Warning; 0 Severe Errors.
   ************* EnergyPlus Terminated--Fatal Error Detected. 0 Warning; 3 Severe Errors; Elapsed Time=00hr 00min  0.12sec
```

Do you know what would cause the entire rest of the IDF to get dumped out instead of just the next line?  I'm on Linux.

**Comment by mbadams5** [comment]
Intriguing. I don't know offhand but likely the index is advancing too far (to the end of the file) in this weird error state. I'll have to recreate locally and debug.

**Comment by mbadams5** [comment]
@Myoldmopar Ok I found the (simple) bug and added a new more descriptive error message for this failure path.

File: src/EnergyPlus/InputProcessing/IdfParser.cc [hunk 1]
```diff
@@ -254,7 +254,7 @@ json IdfParser::parse_idf(std::string const &idf, size_t &index, bool &success,
                 auto found_index = idf.find_first_of('\n', beginning_of_line_index);
                 std::string line;
                 if (found_index != std::string::npos) {
-                    line = idf.substr(beginning_of_line_index, index - beginning_of_line_index);
+                    line = idf.substr(beginning_of_line_index, found_index - beginning_of_line_index);
```
Comment (Myoldmopar) [inline src/EnergyPlus/InputProcessing/IdfParser.cc:257]: 👍

**Comment by Myoldmopar** [comment]
Thanks for getting the extra fix in there @mbadams5, this looks good to go.  Merging.

**Pull request merged**